### PR TITLE
Receiver Detection Sender Reaction policies

### PR DESCRIPTION
**Pull request by kewinrausch**
Introducing the RDSR policies. The 3 policies can be used independently from each other(but the DTCP one has a restriction on avg. bandwidth), allowing users to introduce their own ECN and this DTCP policy, if they want.

The FARE policy allow AE to require at runtime their own avg. bandwidth, which is then handled by the DTCP policy.

Readme under the directory introduce the reader in the policies.

### CODE REVIEW

**Comment by vmaffione** [comment]
````
$ make
make -C /lib/modules/4.1.10-irati/build M=/home/vmaffione/irati/plugins/rdsr-ps/cdrr
make[1]: Entering directory '/home/vmaffione/irati/linux'
  CC [M]  /home/vmaffione/irati/plugins/rdsr-ps/cdrr/dtcp-ps-cdrr.o
/home/vmaffione/irati/plugins/rdsr-ps/cdrr/dtcp-ps-cdrr.c:28:18: fatal error: logs.h: No such file or directory
compilation terminated.
scripts/Makefile.build:258: recipe for target '/home/vmaffione/irati/plugins/rdsr-ps/cdrr/dtcp-ps-cdrr.o' failed
make[2]: *** [/home/vmaffione/irati/plugins/rdsr-ps/cdrr/dtcp-ps-cdrr.o] Error 1
Makefile:1384: recipe for target '_module_/home/vmaffione/irati/plugins/rdsr-ps/cdrr' failed
make[1]: *** [_module_/home/vmaffione/irati/plugins/rdsr-ps/cdrr] Error 2
make[1]: Leaving directory '/home/vmaffione/irati/linux'
Makefile:14: recipe for target 'all' failed
make: *** [all] Error 2
````
And similar errors on the others

**Comment by kewinrausch** [comment]
Yeah, sorry. I was compiling the kernel-mode policies from another directory, so it has the absolute path to the project and not the local one. Now CDRR and ECN should work.

For FARE, you can compile it without having to rebuild all the user space, but in order to do it you need to adjust some path in the Make file: **root** and **rootu** to be more precise, which are the root folder for IRATI repository and the compiled userspace sandbox.

These notes are also written in the main README file, at the end, and the plugin Makefile has the necessary comment to guide you in order to adapt it.

**Comment by edugrasa** [comment]
Did you check the example makefile to build user-space plugins outside of the main src tree?
https://wiki.ict-pristine.eu/wp2/d25/D25-Userspace-Tutorial#Makefile

**Comment by kewinrausch** [comment]
Nope, but I'm adapting it now to align it with the other policies.

**Comment by vmaffione** [comment]
make -C /lib/modules/4.1.10-irati/build M=/home/vmaffione/irati/plugins/rdsr-ps/cdrr
make[1]: Entering directory '/home/vmaffione/irati/linux'
  LD      /home/vmaffione/irati/plugins/rdsr-ps/cdrr/built-in.o
  CC [M]  /home/vmaffione/irati/plugins/rdsr-ps/cdrr/dtcp-ps-cdrr.o
/home/vmaffione/irati/plugins/rdsr-ps/cdrr/dtcp-ps-cdrr.c:103:21: warning: ‘cdrr_time_frame’ defined but not used [-Wunused-variable]
 static unsigned int cdrr_time_frame = CDRR_DEFAULT_TIME_FRAME;
                     ^
  LD [M]  /home/vmaffione/irati/plugins/rdsr-ps/cdrr/cdrr.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC      /home/vmaffione/irati/plugins/rdsr-ps/cdrr/cdrr.mod.o
  LD [M]  /home/vmaffione/irati/plugins/rdsr-ps/cdrr/cdrr.ko
make[1]: Leaving directory '/home/vmaffione/irati/linux'

**Comment by vmaffione** [comment]
please remove "depmod -a" instruction from make clean target

**Comment by vmaffione** [comment]
The Makefile in "fare" directory is just copy pasted. You need to replace the name inside.

**Comment by kewinrausch** [comment]
Fixed everything as you reported.

File: plugins/rdsr-ps/fare/Makefile [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Set INSTALLDIR environment variable to the directory containing
+# the IRATI installation
+
+INSTALLDIR=/users/krausch/pristine/userspace
```
Comment (vmaffione) [inline plugins/rdsr-ps/fare/Makefile]: Remove this line, INSTALLDIR has to be set by the caller, and must not have hardwired paths

**Comment by vmaffione** [comment]
Now it compiles, but I cannot create instances of ecn ps.

````
[   80.458634] rina-rmt-ps-ecn(INFO):  ECNL RMT policy successfully published.
[   80.460122] HERE 5
[   80.460421] rina-ps-factory(ERR): Policy-set instantiation failed
[   80.461229] rina-kipcm(ERR): select-policy-set operation failed
````

I see that ecn_create() fails at 
````
        rmt_cfg = rmt_config_get(rmt);

        if(!rmt_cfg) {
                ecn_queue_set_destroy(data->outqs);
                rkfree(data);
                rkfree(ps);
                return NULL;
        }

````

**Comment by vmaffione** [comment]
The latter is not a problem specific to the ECN policy-set, but it happens with all the RMT policy-sets that call rmt_config_get() in the constructor function, when triggered by a select-policy-set operation.

Tried to use the three policy-sets together, usual three node tests are ok --> ACK

**Pull request merged**